sys/stl-compatibility/include/riot/mutex.hpp

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (C) 2015 Hamburg University of Applied Sciences (HAW)
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup stl-compatibility
+ * @{
+ *
+ * @file    mutex.hpp
+ * @brief   STL mutex drop in replacement
+ * @see     <a href="http://en.cppreference.com/w/cpp/thread/mutex">
+ *            std::mutex, std::lock_guard and std::unique_lock
+ *          </a>
+ *
+ * @author  Raphael Hiesgen <raphael.hiesgen (at) haw-hamburg.de>
+ *
+ * @}
+ */
+
+#ifndef RIOT_MUTEX_HPP
+#define RIOT_MUTEX_HPP
+
+#include "mutex.h"
+
+#include <utility>
+#include <stdexcept>
+#include <system_error>
+
+namespace riot {
+
+class mutex {
+
+ public:
+  using native_handle_type = mutex_t*;
+
+  inline constexpr mutex() noexcept : m_mtx{0, PRIORITY_QUEUE_INIT} {}
+  ~mutex();
+
+  void lock();
+  bool try_lock() noexcept;
+  void unlock() noexcept;
+
+  inline native_handle_type native_handle() { return &m_mtx; }
+
+ private:
+  mutex(const mutex&);
+  mutex& operator=(const mutex&);
+
+  mutex_t m_mtx;
+};
+
+struct defer_lock_t {};
+struct try_to_lock_t {};
+struct adopt_lock_t {};
+
+constexpr defer_lock_t defer_lock = defer_lock_t();
+constexpr try_to_lock_t try_to_lock = try_to_lock_t();
+constexpr adopt_lock_t adopt_lock = adopt_lock_t();
+
+template <class Mutex>
+class lock_guard {
+
+ public:
+  using mutex_type = Mutex;
+
+  inline explicit lock_guard(mutex_type& mtx) : m_mtx(mtx) { m_mtx.lock(); }
+  inline lock_guard(mutex_type& mtx, adopt_lock_t) : m_mtx{mtx} {}
+  inline ~lock_guard() { m_mtx.unlock(); }
+
+ private:
+  mutex_type& m_mtx;
+};
+
+template <class Mutex>
+class unique_lock {
+
+ public:
+  using mutex_type = Mutex;
+
+  inline unique_lock() noexcept : m_mtx{nullptr}, m_owns{false} {}
+  inline explicit unique_lock(mutex_type& mtx) : m_mtx{&mtx}, m_owns{true} {
+    m_mtx->lock();
+  }
+  inline unique_lock(mutex_type& mtx, defer_lock_t) noexcept : m_mtx{&mtx},
+                                                               m_owns{false} {}
+  inline unique_lock(mutex_type& mtx, try_to_lock_t)
+      : m_mtx{&mtx}, m_owns{mtx.try_lock()} {}
+  inline unique_lock(mutex_type& mtx, adopt_lock_t)
+      : m_mtx{&mtx}, m_owns{true} {}
+  inline ~unique_lock() {
+    if (m_owns) {
+      m_mtx->unlock();
+    }
+  }
+  inline unique_lock(unique_lock&& lock) noexcept : m_mtx{lock.m_mtx},
+                                                    m_owns{lock.m_owns} {
+    lock.m_mtx = nullptr;
```

Comment:
You take the configuration from the passed lock and reset its parameters.The [reference](http://en.cppreference.com/w/cpp/thread/unique_lock/unique_lock) uses the following words: `Move constructor. Initializes the unique_lock with the contents of other. Leaves other with no associated mutex.`

But  I am not sure I understand the question.